internal/transport/http2_server.go

```diff
@@ -379,7 +379,7 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 		t.controlBuf.put(&cleanupStream{
 			streamID: streamID,
 			rst:      true,
-			rstCode:  http2.ErrCodeProtocol,
+			rstCode:  http2.ErrCodeInternal,
 			onWrite:  func() {},
 		})
```

Comment:
cc @ejona86/@markdroth to see what other languages are doing / get input.  The cases here are: 1. non-grpc "content-type" header and 2. error decoding `grpc-timeout` header or a `-bin` header.

---

In the case of `!isGRPC`, I think the right response is an HTTP status (_maybe_ 406 Not Acceptable(??)) in a HTTP/2 HEADERS frame with END_STREAM set.

In the `headerError` case, the client is apparently speaking gRPC, so we could return a gRPC status (`INTERNAL` or `UNKNOWN`).

In both cases, we would RST_STREAM with NO_ERROR after returning the trailers unless the client has sent an END_STREAM in the initial headers (unexpected for gRPC, but certainly possible in HTTP/2).

I'm not sure whether the control buf is currently able to do that for non-"accepted"/gRPC streams, however -- you'd need to do some digging in there to figure out how it works.

Reply:
Looks like you didn't actually tag @ejona86 there.

@yashykt should be able to tell you what the C-core behaviors are in these cases.

Reply:
Java behavior:
1. Bad content-type. HTTP 415 with a text/plain error message. Also a gRPC status of INTERNAL with same error message
2. Bad grpc-timeout or bad -bin. Looks like Java throws an exception (from different spots) that should eventually turn into a RST_STREAM(INTERNAL). Not the greatest approach, but not the worst result

Reply:
Yes, 415 is definitely better than 406.

> Looks like Java throws an exception

What _should_ we do?  Do you think a graceful gRPC status would be ideal?

Reply:
> What should we do?

I don't know! I side-stepped that! :smile:

I do think a graceful gRPC status (INTERNAL seems fair) would be ideal. We shouldn't rely on RST_STREAM often, in my mind, and there's little reason other than laziness to use RST_STREAM at the beginning of the call. In this case a useful error message along with a gRPC status would help a ton with debugging if these code flows were ever hit. But this case isn't also all that common, so a simple RST_STREAM(INTERNAL) isn't _that_ bad.

Reply:
Core seems to just ignore it after logging a message -
1. https://github.com/grpc/grpc/blob/8bdbb74425d6ac13ea7823be41e9297fddd629aa/src/core/ext/filters/http/server/http_server_filter.cc#L250
2. https://github.com/grpc/grpc/blob/8bdbb74425d6ac13ea7823be41e9297fddd629aa/src/core/ext/transport/chttp2/transport/parsing.cc#L430